### ISSUE TITLE
Extension Rewrites Host Header

### ISSUE THREAD

**Issue by amiller-si** [NetSPI/AWSSigner#33]
The signer has changed header handling behavior that is negatively impacting some of my testing. Specifically, when I make a request in Repeater that has a Host header that is different than the Target set, the signer rewrites the Host header to be equivalent to the Repeater Target. With my proxy setup, this is a breaking change and I had to grab an old version of the signer in order to continue my work. Please consider adding a setting to modify this behavior or turn it off.

Ex. 

Initial request in repeater:
Target: localhost:12345
Host: vpce-blahblah.com

Result, as seen in Flow (logging extension):
Target: localhost:12345
Host: localhost:12345

Also, I’m unclear if it uses the Host or the Target to sign, but I would want the Host to be the signed header.

**Comment by jakekarnes42** [comment]
Thank you for the heads up on this. If you had to revert to an older version of the extension to make this work, that sounds like a problem. I'm sorry that was your experience here. I'll take a closer look and get back to you when I have a better understanding of the root cause.

**Comment by jakekarnes42** [comment]
I think I see the issue here. The [SigV4 signature](https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html), must have Host as one of the signed headers. The v2 of the signer extension delegates to the AWS Java SDK to do the heavy lifting. The [SDK pulls the Host header value out of the URL](https://github.com/aws/aws-sdk-java-v2/blob/06878b7fc531e1515d3a1e98bcb62a60c08ec785/core/auth/src/main/java/software/amazon/awssdk/auth/signer/internal/AbstractAws4Signer.java#L388) when signing a request. We populate that URL based on the request's target URL. V1 of the signer extension builds the signature without the SDK and uses the Host header as expected. 

So long story short: you're right. It's signing the request and overwriting the Host header based on the target URL. And I could see how this would break your proxy setup. The goal of V2 is to maintain compatibility with V1 so I'd like to get closer to the original behavior. 

I think I have a fix, but it would be great if you could validate that it works with your proxy setup. If I can build and upload a release candidate JAR here, would you be willing to try it out for me?

**Comment by amiller-si** [comment]
Certainly. Thanks so much for the prompt review of this.

**Comment by jakekarnes42** [comment]
Thanks for helping out with this. Please test out this JAR with your proxy setup when you get the chance. 

If it's working as expected, your original repeater setup should work now. The request should be signed with the vpce-blahblah.com as the host name, but the request should still be sent to localhost:12345 to be proxied through. 

Please let me know if you run into any trouble though. If it's working, I'll push this out as a fix. 
[awssigner-2.0.2-rc-host-header-fix.jar.zip](https://github.com/NetSPI/AWSSigner/files/8643598/awssigner-2.0.2-rc-host-header-fix.jar.zip)

**Comment by amiller-si** [comment]
There's still one issue. It is tacking on the port to the host header, which again causes the signature verification to fail as far as I can tell. For some reason, I couldn't see this change in Flow (doesn't appear to capture responses for signed requests when signer is on, not really an issue for me but interesting) but it was visible in Logger. 

Initial request in Repeater:
Target: localhost:12345
Host: vpce-blahblah.com

Result, as seen in Logger:
Target: localhost:12345
Host: vpce-blahblah.com:12345

**Comment by jakekarnes42** [comment]
Oof 🤦‍♂️

Sorry about that. But this is exactly why I wanted you try it out. So thanks again for your patience and for working with me. 

I think I've got it this time. It should handle the cases where the port is omitted or included in either the URL or the Host header:
[awssigner-2.0.2-rc-host-header-fix-2.jar.zip](https://github.com/NetSPI/AWSSigner/files/8643721/awssigner-2.0.2-rc-host-header-fix-2.jar.zip)

**Comment by amiller-si** [comment]
It's no trouble. I'm not hitting the port issue anymore, so I think it's good to go.

**Comment by jakekarnes42** [comment]
The fix has been merged in with version 2.0.3.
